### DISCUSSION TITLE
How are the hot swap with HS_NUB suppose to work

### DISCUSSION THREAD

**Discussion by Looooopy** [joshreve/dactyl-keyboard#66] [Q&A] [answered]
Problem 1: 
When I tried to print it and the tested the fitting if felt like the notch was oriented wrong according to how the switch goes in.
The Switch has levers which is position on top and bottom (Picture 1)
The Notch is placed on right and left side (Picture 2).

Problem 2:
I am not certain if it is dependent on the notch but when I push down the socket it fits on top (Picture 3) but the bottom (Picture 4) is like 1mm off.

Question 1: Is the notch placed correctly or can I tweak the position of the notch on top and bottom instead of left and right?
Question 2: Is it possible to tweak the depth of the socket to make it deeper.

Picture 1, switch (problem 1)
![image](https://user-images.githubusercontent.com/16755871/143526000-4dcd2b77-4ffd-4638-a9f8-65b9c673f74b.png)

Picture 2, socket with a notch and hot swap (problem 1)
![image](https://user-images.githubusercontent.com/16755871/143525773-f4cb884a-2610-4ae2-b3b8-da4215687d2e.png)

Picture 3, socket and switch top (problem 2)
![image](https://user-images.githubusercontent.com/16755871/143527269-218cbb1a-754b-4ca4-9aba-f28ae7e775fd.png)

Picture 4, socket and switch bottom (problem 2)
![image](https://user-images.githubusercontent.com/16755871/143527193-a14161f8-7344-44bc-b400-d8358db87bf1.png)

**Comment by joshreve** [comment, marked as answer]
Looks like you are referring to the nubs not the notches.  I'll be honest, I don't like the nubs and have never used them.  In all previous images I can find they are oriented left to right with the switch clips top and bottom.  I think the nubs go into the big square undercut on the side of the switch?  Either way, I can't find a nub and hotswap image online, probably due to this conflict.  I would switch to the notches.  Images I can find either use notches or a bumped out version of the notch intended for the clips when using a hotswap plate.  I would suggest using HS_NOTCH and tuning the notch undercut to get a good clip action.  I took my best guess at a undercut thickness but it is based on my printer and settings so you may need to tweak it a bit to get the best results.

If you are running "HS_NOTCH" and getting this let me know.  I did confirm I get the right geometry for mine, but there is a possibility of a bug if you are running a different workflow.  I am tempted to just remove the nubs as I think they are inferior to actually using real switch clip geometries.  I think they exist due to the large variations 3d printers have, especially back when the initial version of the keyboard was created.

**Reply by Looooopy** [reply]
Yeah you are right I was using HS_NUB and I think I selected that because of the comment on HS_UNDERCUT and HS_NOTCH
`Does not generate properly.  Hot swap step needs to be modified.` 

* Sounds like removing HS_NUB would be a sane thing to do.
* Change the comment on HS_NOTCH to something else so it does not feels like the user select something that is totally broken :)

I have now tried the HS_NOTCH and it looks nicer - I will se how it works when I do a print.

Follow up question:
I have done some digging on the my "Question 2" above and found the setting in config "plate_offset" and that config is override to always be '0' when hot swap is active.

> is it because we don't handle it correctly on the models later on or is it a mistake?
https://github.com/joshreve/dactyl-keyboard/blob/master/src/dactyl_manuform.py#L118
- Should a new config be introduced say for example 'hs_plate_offset' that is set to 0 by default that overrides 'plate_offset' instead of hardcoding an override?
- I might not need it now when I switched to HS_NOTCH but I will comment that line and see how thing behaves if I need some more space for my switch so it  makes contact on the bezel of the switch (Picture 4, above).

Thanks for all support you are giving me!

**Reply by joshreve** [reply]
No problem, I appreciate the debug help.  There is no way to test all configs, so I have to rely on people to communicate issues.  I appreciate the time you've spend clearly describing the issues you encounter, it makes it easy to correct them.

On this one, the hardcoded 0 shouldn't be there.  You can comment it out and try.  The imported plate has limitations and isn't as adaptable as the native constructed geometry.  I plan to convert it at some point, but it isn't at the top of the list.  You can use the setting from the config file, just double check you don't create gaps as that hasn't been thoroughly tested.  If you try it and there is an issue let me know, otherwise, it should work fine with small offsets after commenting that line out.

**Comment by bosd** [comment]
Wish, I found this info earlier.

Just received my 3d prints, and have the same issues.
Could'nt overcome the cadquery issues on my system (and in the docker) so, decided to print the stl files from the
[Alpine project](
https://github.com/nol00p/Alpine).

Sadly, the switches do not firt. (at first I was guessing this would be because of my material choice, Nylon)

Here are the results.
![IMG_20220524_105657_detail](https://user-images.githubusercontent.com/11499387/170005276-bbaf19ad-2bb8-4aa0-b9c3-6709504f6cd3.jpg)

Sadly, this print is not usable. :cry: 

corners are not fitting as well.
![IMG_20220524_093716_detail2](https://user-images.githubusercontent.com/11499387/170006641-c881468c-8490-49ef-ba00-095a6d65a1ab.jpg)

**Comment by bosd** [comment]
The printshop also sent me the misprints. In this one they ruined one of the keyholes.
![image](https://user-images.githubusercontent.com/11499387/170015825-b900b090-929d-41c1-8e76-c54384fd7d0c.png)

So I put one key in, rotated 90 degrees. And we have a winner :partying_face: :clap: 
The fit is perfect. :tada: 
![image](https://user-images.githubusercontent.com/11499387/170016400-10d3dfc4-ae3a-488c-958a-67f4d72c7457.png)



If I am really nitpicking:
![image](https://user-images.githubusercontent.com/11499387/170010853-b7af6c57-7111-4ce8-9587-c12c470f644a.png)
Click to see 
<details>
I would make the nothes 0.1mm smaller as it seems the plastic is stretched at most.
Compare the picture where I marked it with an red arrow (Comfortable stretch) against the picture when I rotated the key 90 degr, Blue arrow.

Would not say that this is due the fact the notch is to big. I think its more like the notch is not aligned properly in the Z-plane.
Yep, upon closer inspection, the Z of the notch is too high. Meaning the notch goes to deeply to the bottom. It should stop a bit earlier. So the part marked with the red/blue arrow should snap back more to its natural position. 





</details>
Eventually in the design I would love to have the notches and the baseplate.
(With baseplate I mean the part that holds the black hot swap holder)

Conclusion: If the baseplate could be rotated 90 degrees. we have a perfect fit)

**Comment by bosd** [comment]
![image](https://user-images.githubusercontent.com/11499387/170468078-bff7266e-0a0e-4abc-aa40-e0d22fab8734.png)

source: https://www.keychatter.com/wp-content/uploads/2018/08/kailh-box-switch-specifications-NEW.jpg

**Reply by Looooopy** [reply]
It looks like they have some serious under extrusion on their printer as well the gaps in the plastic tells me that.
I know because I have struggled with it for a while with a stock printer CR10S :)

I think if I would order from a shop they should be better on tuning printers otherwise they would not sell anything, did you order from a professional shop or is it a friend who did it for you?

**Reply by bosd** [reply]
It's from treatstock.com Mago
Paid €123 for that job. 

I'm consedering placing an with jlcb. 
As I still need to order the mxledbit for my build. It is a good chance to test their service.
(Considerably cheaper)

Before I place that order, I'd like to fix the design on the points mentioned above. But I don't know how to do that..

**Reply by bosd** [reply]
The oled holders are even worse!

![IMG_20220524_093911.jpg](https://user-images.githubusercontent.com/11499387/170769651-f0a3f241-0980-401a-8f6e-dfe66ccc26a3.jpg)

**Reply by Looooopy** [reply]
i would like to say you have a right to reclaim your money back because the poor job they did ....
